cmd/csaf_checker/processor.go

```diff
@@ -1084,83 +1077,157 @@ func (p *processor) checkProviderMetadata(domain string) error {
 	}
 	if res.StatusCode != http.StatusOK {
-		p.badSecurity.error("Fetching %s failed. Status code %d (%s)",
+		return fmt.Sprintf("Fetching %s failed. Status code %d (%s)",
 			u, res.StatusCode, res.Status)
-		return errContinue
 	}
 	defer res.Body.Close()
 	// Compare checksums to already read provider-metadata.json.
 	h := sha256.New()
 	if _, err := io.Copy(h, res.Body); err != nil {
-		p.badSecurity.error("Reading %s failed: %v", u, err)
-		return errContinue
+		return fmt.Sprintf("Reading %s failed: %v", u, err)
 	}
 
 	if !bytes.Equal(h.Sum(nil), p.pmd256) {
-		p.badSecurity.error("Content of %s from security.txt is not "+
+		return fmt.Sprintf("Content of %s from security.txt is not "+
 			"identical to .well-known/csaf/provider-metadata.json", u)
 	}
+	return ""
+}
+
+// checkDNS checks if the "csaf.data.security.domain.tld" DNS record is available
+// and serves the "provider-metadata.json".
+// It returns an empty string if all checks are passed, otherwise the errormessage.
+func (p *processor) checkDNS(domain string) string {
+
+	client := p.httpClient()
+	path := "https://csaf.data.security." + domain
+	res, err := client.Get(path)
+	if err != nil {
+		return fmt.Sprintf("Fetching %s failed: %v", path, err)
+	}
+	if res.StatusCode != http.StatusOK {
+		return fmt.Sprintf("Fetching %s failed. Status code %d (%s)",
+			path, res.StatusCode, res.Status)
+
+	}
+	hash := sha256.New()
+	defer res.Body.Close()
+	content, err := io.ReadAll(res.Body)
+	if err != nil {
+		return fmt.Sprintf("Error while reading the response from %s", path)
+	}
+	hash.Write(content)
+	if !bytes.Equal(hash.Sum(nil), p.pmd256) {
+		return fmt.Sprintf("%s does not serve the same provider-metadata.json as previously found", path)
+	}
+	return ""
+}
+
+// checkWellknownMetadataReporter checks if the provider-metadata.json file is
+// available under the /.well-known/csaf/ directory. Returns the errormessage if
+// an error was encountered, or an empty string otherwise
+func (p *processor) checkWellknown(domain string) string {
+
+	client := p.httpClient()
+	path := "https://" + domain + "/.well-known/csaf/provider-metadata.json"
+
+	res, err := client.Get(path)
+	if err != nil {
+		return fmt.Sprintf("Fetching %s failed: %v", path, err)
+	}
+	if res.StatusCode != http.StatusOK {
+		return fmt.Sprintf("Fetching %s failed. Status code %d (%s)",
+			path, res.StatusCode, res.Status)
+	}
+	return ""
+}
+
+// checkWellknownSecurityDNS checks if the provider-metadata.json file is
+// available under the /.well-known/csaf/ directory. Then it checks the security.txt file
+// by making HTTP request to fetch it.
+// After that it checks the existence of the CSAF field in the file content and tries to fetch
+// the value of this field.
+// Finally it checks if the "csaf.data.security.domain.tld" DNS record is available
+// and serves the "provider-metadata.json".
+// If all three checks fail, errors are given,
+// otherwise warnings for all failed checks.
```

Comment:
The DNS check should be just "informational" if a PMD was found by another means.